pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/DeletePatch.java

```diff
@@ -0,0 +1,83 @@
+package org.broadinstitute.dsm.model.patch;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.broadinstitute.dsm.db.dao.DeletedObjectDao;
+import org.broadinstitute.dsm.db.dao.ddp.instance.DDPInstanceDao;
+import org.broadinstitute.dsm.db.dto.ddp.instance.DDPInstanceDto;
+import org.broadinstitute.dsm.exception.DsmInternalError;
+import org.broadinstitute.dsm.exception.UnsafeDeleteError;
+import org.broadinstitute.dsm.model.NameValue;
+import org.broadinstitute.dsm.util.ElasticSearchDeleteUtil;
+import org.broadinstitute.dsm.util.NotificationUtil;
+
+@Slf4j
+public class DeletePatch extends ExistingRecordPatch {
+    DeleteType deleteType;
+
+    public DeletePatch(Patch patch, NotificationUtil notificationUtil, DeleteType deleteType) {
+        super(patch, notificationUtil);
+        this.dbElementBuilder = new DefaultDBElementBuilder();
+        this.deleteType = deleteType;
+    }
+
+    @Override
+    public Object doPatch() {
+        //todo this should get changed after 1209 is merged
+        if (!isSafeToDelete() || !patch.isDeleteAnyway()) {
+            throw new UnsafeDeleteError("This object is used in a clinical order");
+        }
+        //recursively start by creating patches for "children" of the current
+        DeletePatchFactory.deleteChildrenFields(this.patch, this.getNotificationUtil());
+        if (isNameValuePairs()) {
+            throw new DsmInternalError(String.format("Patch is invalid for NameValues {}, delete patch should have only one name value",
+                    patch.getNameValues()));
+        }
+        return patchNameValuePair();
+    }
+
+    @Override
+    public Object patchNameValuePair() {
+        Optional<Object> maybeNameValue = processSingleNameValue();
+        return maybeNameValue.orElse(null);
+    }
+
+    @Override
+    Object handleSingleNameValue() {
+        List<NameValue> nameValues = new ArrayList<>();
+        if (DeletedObjectDao.deletePatch(patch.getId(), patch.getUser(), dbElement, deleteType)) {
+            DDPInstanceDto ddpInstanceDto = new DDPInstanceDao().getDDPInstanceByInstanceName(patch.getRealm()).orElseThrow();
+            try {
+                ElasticSearchDeleteUtil.deleteFromIndexById(patch.getDdpParticipantId(), Integer.parseInt(patch.getId()), ddpInstanceDto, deleteType);
+            } catch (Exception e) {
+                throw new DsmInternalError("Unable to delete from ES for patch with id "+patch.getId() +" for NameValue: "+patch.getNameValue().toString(),e);
```

Comment:
Is there something we can log here to help us recover from a successful db delete but a failed ES delete?  If we see this exception thrown, how would we go about synchronizing the state of ES?  A full export?

Reply:
Actually that is a great concern of mine too, I made a ticket to be able to export only one participant data by a  `ddp_participant_id`, so maybe we can think of prioritizing that ticket? 

One other solution may be to roll back a db delete if ES delete does not work. I don't think it would be bad to consider that. What do you think about that?

Reply:
Happy to up-prioritize the export of a single participant!

For rollback, we want database transactions to be as short as possible to avoid db contention, and avoid external network calls as much as possible while the transaction is open.  Otherwise, when those external calls slow down, database contention can become a big issue quickly.  So let's leave this as-is and work on the single participant export in a separate ticket as you suggest.